File: campus-navigate-app/expo-app/src/components/map/actor/actorDirectionModel.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';

import {
  actorDirectionFromHeading,
  actorDirectionWithHysteresis,
} from './actorDirectionModel';

test('maps the map heading convention to four upright actor directions', () => {
  assert.equal(actorDirectionFromHeading(0), 'right');
  assert.equal(actorDirectionFromHeading(44), 'right');
  assert.equal(actorDirectionFromHeading(45), 'down');
  assert.equal(actorDirectionFromHeading(90), 'down');
  assert.equal(actorDirectionFromHeading(135), 'left');
  assert.equal(actorDirectionFromHeading(180), 'left');
  assert.equal(actorDirectionFromHeading(225), 'up');
  assert.equal(actorDirectionFromHeading(270), 'up');
  assert.equal(actorDirectionFromHeading(315), 'right');
});

test('normalizes negative and wrapped headings', () => {
  assert.equal(actorDirectionFromHeading(-90), 'up');
  assert.equal(actorDirectionFromHeading(-180), 'left');
  assert.equal(actorDirectionFromHeading(450), 'down');
});

test('holds the current actor direction inside a ten-degree boundary buffer', () => {
  assert.equal(
    actorDirectionWithHysteresis({
      currentDirection: 'right',
      headingDegrees: 54,
    }),
    'right',
  );
  assert.equal(
    actorDirectionWithHysteresis({
      currentDirection: 'right',
      headingDegrees: 56,
    }),
    'down',
  );
  assert.equal(
    actorDirectionWithHysteresis({
      currentDirection: 'down',
      headingDegrees: 36,
    }),
    'down',
  );
  assert.equal(
    actorDirectionWithHysteresis({
      currentDirection: 'down',
      headingDegrees: 34,
    }),
    'right',
  );
});

File: campus-navigate-app/flutter_app/test/ui/map/models/actor_direction_model_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:indoor_navigation/ui/map/models/actor_direction_model.dart';
import 'package:indoor_navigation/ui/map/models/actor_models.dart';

void main() {
  test('maps the map heading convention to four upright actor directions', () {
    expect(actorDirectionFromHeading(0), ActorDirection.right);
    expect(actorDirectionFromHeading(44), ActorDirection.right);
    expect(actorDirectionFromHeading(45), ActorDirection.down);
    expect(actorDirectionFromHeading(90), ActorDirection.down);
    expect(actorDirectionFromHeading(135), ActorDirection.left);
    expect(actorDirectionFromHeading(180), ActorDirection.left);
    expect(actorDirectionFromHeading(225), ActorDirection.up);
    expect(actorDirectionFromHeading(270), ActorDirection.up);
    expect(actorDirectionFromHeading(315), ActorDirection.right);
  });

  test('normalizes negative and wrapped headings', () {
    expect(actorDirectionFromHeading(-90), ActorDirection.up);
    expect(actorDirectionFromHeading(-180), ActorDirection.left);
    expect(actorDirectionFromHeading(450), ActorDirection.down);
  });

  test('holds direction inside the exact ten-degree boundary buffer', () {
    expect(
      actorDirectionWithHysteresis(
        currentDirection: ActorDirection.right,
        headingDegrees: 54,
      ),
      ActorDirection.right,
    );
    expect(
      actorDirectionWithHysteresis(
        currentDirection: ActorDirection.right,
        headingDegrees: 56,
      ),
      ActorDirection.down,
    );
    expect(
      actorDirectionWithHysteresis(
        currentDirection: ActorDirection.down,
        headingDegrees: 36,
      ),
      ActorDirection.down,
    );
    expect(
      actorDirectionWithHysteresis(
        currentDirection: ActorDirection.down,
        headingDegrees: 34,
      ),
      ActorDirection.right,
    );
  });
}
